internal/providers/github/properties.go

```diff
@@ -0,0 +1,167 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package github
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/stacklok/minder/internal/entities/properties"
+	minderv1 "github.com/stacklok/minder/pkg/api/protobuf/go/minder/v1"
+)
+
+const (
+	// RepoPropertyName represents the github repository name
+	RepoPropertyName = "github/repo_name"
+	// RepoPropertyOwner represents the github repository owner
+	RepoPropertyOwner = "github/repo_owner"
+	// RepoPropertyDeployURL represents the github repository deployment URL
+	RepoPropertyDeployURL = "github/deploy_url"
+	// RepoPropertyCloneURL represents the github repository clone URL
+	RepoPropertyCloneURL = "github/clone_url"
+	// RepoPropertyDefaultBranch represents the github repository default branch
+	RepoPropertyDefaultBranch = "github/default_branch"
+)
+
+type propertyWrapper func(ctx context.Context, ghCli *GitHub, name string) (map[string]any, error)
+
+type propertyOrigin struct {
+	keys    []string
+	wrapper propertyWrapper
+}
+
+type propertyFetcher struct {
+	ghCli *GitHub
+
+	propertyOrigins []propertyOrigin
+}
+
+var repoPropertyDefinitions = []propertyOrigin{
+	{
+		keys: []string{
+			// general entity
+			properties.PropertyName,
+			properties.PropertyUpstreamID,
+			// general repo
+			properties.RepoPropertyIsPrivate,
+			properties.RepoPropertyIsArchived,
+			properties.RepoPropertyIsFork,
+			// github-specific
+			RepoPropertyName,
+			RepoPropertyOwner,
+			RepoPropertyDeployURL,
+			RepoPropertyCloneURL,
+			RepoPropertyDefaultBranch,
+		},
+		wrapper: getRepoWrapper,
+	},
+}
+
+func newRepoPropertyFetcher(ghCli *GitHub) *propertyFetcher {
+	return &propertyFetcher{
+		ghCli:           ghCli,
+		propertyOrigins: repoPropertyDefinitions,
+	}
+}
+
+func newPropertyFetcher(ghCli *GitHub, entType minderv1.Entity) *propertyFetcher {
+	if entType != minderv1.Entity_ENTITY_REPOSITORIES {
+		return nil
+	}
+
+	return newRepoPropertyFetcher(ghCli)
+}
+
+func getRepoWrapper(ctx context.Context, ghCli *GitHub, name string) (map[string]any, error) {
+	// TODO: this should be a provider interface, even if private
+	slice := strings.Split(name, "/")
+	if len(slice) != 2 {
+		return nil, errors.New("invalid name")
+	}
+
+	repo, err := ghCli.GetRepository(ctx, slice[0], slice[1])
+	if err != nil {
+		return nil, err
+	}
+
+	repoProps := map[string]any{
+		// general entity
+		properties.PropertyName:       fmt.Sprintf("%s/%s", repo.GetOwner().GetLogin(), repo.GetName()),
+		properties.PropertyUpstreamID: repo.GetID(),
```

Comment:
I'm not sure about this, but I'd like to get your opinion: Should we serialize the upstream ID as a string? This would probably change from provider to provider. Some might use serial integers, while others might use a UUID or another sort of random identifier. What do you think? Maybe this is irrelevant for now, but I wanted to get your thoughts on this.

Reply:
Ah this is actually a great idea, the generic ID should be a string and we should have a github-specific property as an int.